docs/topics/spot-vms.md

```diff
@@ -0,0 +1,52 @@
+# Spot Virtual Machines
+
+[Azure Spot Virtual Machines](https://azure.microsoft.com/en-gb/pricing/spot/) allow users to reduce the costs of their
+compute resources by utilising Azure's spare capacity for a lower price.
+
+With this lower cost, comes the risk of preemption.
+When capacity within a particular Availability Zone is increased,
+Azure may need to reclaim Spot Virtual Machines to satisfy the demand on their data centres.
+
+## When should I use Spot Virtual Machines?
+
+Spot Virtual Machines are ideal for workloads that can be interrupted.
+For example, short jobs or stateless services that can be rescheduled quickly,
+without data loss, and resume operation with limited degradation to a service.
+
+## How do I use Spot Virtual Machines?
+
+**Note**: This feature is only available on Machines at present and not on the
+experimental MachinePools.
+
+To enable a Machine to be backed by a Spot Virtual Machine, add `spotMarketOptions`
+to your `AzureMachineTemplate`:
+
+```yaml
+apiVersion: exp.infrastructure.cluster.x-k8s.io/v1alpha3
+kind: AzureMachineTemplate
+metadata:
+  name: capz-mp-0
+spec:
+  location: westus2
+  template:
+    osDisk:
+      diskSizeGB: 30
+      managedDisk:
+        storageAccountType: Premium_LRS
+      osType: Linux
+    sshPublicKey: ${YOUR_SSH_PUB_KEY}
+    vmSize: Standard_D2s_v3
+    spotMarketOptions: {}
```

Comment:
Wouldn't it make more sense to have spotMarketOptions.Enabled set to true or false? Having an empty field to "enable" a feature seems a little strange... I realize this is an API question but looking at the example is what made me realize that was the way of enabling it without maxPrice

Reply:
This was inspired by the `emptyDir: {}` volume from core Kubernetes. This was documented in the proposal to cluster-api and the same approach was intended to be taken for the AWS provider too. If you think we should change this, we should have a discussion about changing the proposal and seeing what the AWS folks think as well, ideally there would be consistency between the two providers IMO

Reply:
@devigned @detiber any thoughts on this one?

Reply:
That looks really odd.

I'm not sure what is worse, having to set enabled and possibly a max price, which seem redundant, or having to specify an empty.

Reply:
It seems to be part of the idiom: https://kubernetes.io/docs/concepts/storage/volumes/#example-pod. Anyone have any better ideas?

Reply:
the `emptyDir` volume feels slightly different to me because it does point to something being empty. SpotVMs is a feature you can enable/disable, with optional options, like NTP https://github.com/kubernetes-sigs/cluster-api/blob/master/bootstrap/kubeadm/api/v1alpha2/kubeadmbootstrapconfig_types.go#L197 so I would prefer something like:

```
// Spot defines the options relevant to running the Machine on Spot VMs
type Spot struct {
	// MaxPrice defines the maximum price the user is willing to pay for Spot VM instances
	// +optional
	// +kubebuilder:validation:Type=number
	MaxPrice *string `json:"maxPrice,omitempty"`

	// Enabled specifies whether Spot should be enabled
	// +optional
	Enabled *bool `json:"enabled,omitempty"`
}
```

Would be good to get CAPA folks to chime in as well since we want to be consistent across providers.

Reply:
@kubernetes-sigs/cluster-api-provider-aws-maintainers you all have any thoughts on how CAPA is thinking about approaching spot pricing? I did a quick search through the issues and didn't see anything.

Reply:
I don't think the empty struct vs nil bothers me much if at all, since in the default case a user just wouldn't specify the field and would not get a Spot instance, only when they explicitly initialize/set the field do they get a Spot instance.

To me it would likely feel a bit odder if you were to explicitly configure the optional fields for the SpotVMOptions but omit setting Enabled to true and end up with a non-spot VM.

Reply:
Thanks for chiming it @detiber. I'm okay with going with empty options if that's the general consensus and since that was the original agreed-upon design. @devigned wdyt?

Reply:
👍 Let's do this!

@JoelSpeed, do you want to continue this PR? Since it's been a minute, just wanted to see if you were still interested in carrying this over the finish line. If you don't have time, I'm happy to help.

Reply:
Yep, happy to continue, will rebase tomorrow